### ISSUE TITLE
[FR] Unnecessary type casting: new pattern

### ISSUE THREAD

**Issue by orklah** [kalessil/phpinspectionsea#1215]
Description :
--

This code is not detected by the "Unnecessary type casting" inspection:
```php
class a{
    /**
     * @return int
     */
    function return_int(){
        return 1;
    }
    function run()
    {
        $a = (int)$this->return_int();
    }
}
```

If it's a function instead of a method it works great.

Thanks

**Comment by orklah** [comment]
I had an other case but I'm not sure it's worth an inspection or if it's this inspection that should do it...

I came across this code: 
```php
is_numeric((int)$a);
```

Either $a is already a numeric and the cast is unnecessary or $a is not a numeric and the cast will make it a numeric by force:
https://3v4l.org/660gl

In either case, the statement returns true... (except maybe if $a is an object, this will throw some kind of error)

**Comment by orklah** [comment]
I guess while we're at it

```php
is_numeric(return_strictly_int_or_float());
```

Could be flagged as always true

**Comment by kalessil** [comment]
- `(int)$this->return_int()`:  in this case skipped, as no implicit return type specified. PhpDoc can notbe trusted
- `is_numeric((int)...)`, `is_numeric((float)...)`, `is_numeric(call_with_implicit_return_type())`: nice, expressions with (possibly) constant result.

**Comment by orklah** [comment]
> * `(int)$this->return_int()`:  in this case skipped, as no implicit return type specified. PhpDoc can not be trusted

For consistency, it would be great to have a similar behaviour between functions and methods.

```php
/**
 * @return int
 */
function function_return_int(){
    return 1;
}
$a = (int)function_return_int();
```

This triggers unnecessary type casting but the code in the first post does not.

We should either drop the function support or implement method support.

**Comment by kalessil** [comment]
Re-checked: functions/methods are handled equally (int the example with `function_return_int`) return type is missing.

**Comment by kalessil** [comment]
Implemented `is_numeric((int)...), is_numeric((float)...), is_numeric(call_with_implicit_return_type())` part.
